starfish/util/synthesize.py

```diff
@@ -0,0 +1,202 @@
+        'N_flour': 200,
+        # mean number of photons per flourophore - depends on exposure time, bleaching rate of dye
+        'N_photons_per_flour': 50,
+        # mean number of background photons per pixel - depends on tissue clearing and autoflourescence
+        'N_photon_background': 1000,
+        # quantum efficiency of the camera detector units number of electrons per photon
+        'detection_efficiency': .25,
+        # camera read noise per pixel in units electrons
+        'N_background_electrons': 1,
+        # number of RNA puncta; keep this low to reduce overlap probability
+        'N_spots': 20,
+        # height and width of image in pixel units
+        'N_size': X,
+        # standard devitation of gaussian in pixel units
+        'psf': 2,
+        # dynamic range of camera sensor 37,000 assuming a 16-bit AD converter
+        'graylevel': 37000.0 / 2 ** 16,
+        # 16-bit AD converter
+        'bits': 16
+    }
+
+    codebook = graham_sloane_codes(p['N_barcode'])
+
+    def generate_spot(p):
+        position = rand(2)
+        gene = random.choice(range(len(codebook)))
+        barcode = array(codebook[gene])
+        photons = [poisson(p['N_photons_per_flour']) * poisson(p['N_flour']) * b for b in barcode]
+        return DataFrame({'position': [position], 'barcode': [barcode], 'photons': [photons], 'gene': gene})
+
+    # right now there is no jitter on x-y positions of the spots, we might want to make it a vector
+    spots = concat([generate_spot(p) for _ in range(p['N_spots'])])  # type: ignore
+
+    image = zeros((p['N_barcode'], p['N_size'], p['N_size'],))
+
+    for s in spots.itertuples():
+        image[:, int(p['N_size'] * s.position[0]), int(p['N_size'] * s.position[1])] = s.photons
+
+    image_with_background = image + poisson(p['N_photon_background'], size=image.shape)
+    filtered = array([gaussian(im, p['psf']) for im in image_with_background])
+    filtered = filtered * p['detection_efficiency'] + normal(scale=p['N_background_electrons'], size=filtered.shape)
+    signal = np.array([(x / p['graylevel']).astype(int).clip(0, 2 ** p['bits']) for x in filtered])
+
+    def select_uint_dtype(array):
+        """choose appropriate dtype based on values of an array"""
+        max_val = np.max(array)
+        for dtype in [np.uint8, np.uint16, np.uint32, np.uint64]:
+            if max_val <= dtype(-1):
+                return array.astype(dtype)
+        raise ValueError('value exceeds dynamic range of largest numpy type')
+
+    corrected_signal = select_uint_dtype(signal)
+    rescaled_signal: np.ndarray = rescale_intensity(corrected_signal)
+
+    # set up the tile set
+    image_data = TileSet(
+        {Coordinates.X, Coordinates.Y, Indices.HYB, Indices.CH, Indices.Z},
+        {
+            Indices.HYB: NUM_HYB,
+            Indices.CH: NUM_CH,
+            Indices.Z: NUM_Z,
+        },
+        default_tile_shape=(Y, X),
+    )
+
+    # fill the TileSet
```

Comment:
Can this use `synthetic_stack`?

Reply:
Yea, I think it can. I'll give a refactor a go.

Reply:
I think this should work, but I'd like to do this in a future PR. I think it's more important to finish the IntensityTable / codebook integration. I've addressed the other comments.

Reply:
Sounds good.